data/transactions/logic/TEAL_opcodes.md

```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+| 0 | BN254g1 | G1 of the BN254 curve. Points encoded as 32 byte X following by 32 byte Y |
+| 1 | BN254g2 | G2 of the BN254 curve. Points encoded as 64 byte X following by 64 byte Y |
+| 2 | BLS12_381g1 | G1 of the BLS 12-381 curve. Points encoded as 48 byte X following by 48 byte Y |
+| 3 | BLS12_381g2 | G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 48 byte Y |
+
+
+A and B are curve points in affine representation: X concatenated with Y. Fails if A or B is not in G. Does _not_ check if A and B are in the main prime-order subgroup.
+
+## ec_scalar_mul g
+
+- Opcode: 0xe1 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve point A and scalar B, return the curve point BA
+- **Cost**:  BN254g1=2200 BN254g2=4460 BLS12_381g1=3640 BLS12_381g2=8530
+- Availability: v9
+
+A is a curve point encoded and checked as described in `ec_add`. Scalar B is interpreted as a big-endian unsigned integer. Fails if B exceeds 32 bytes.
+
+## ec_pairing_check g
+
+- Opcode: 0xe2 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64
+- 1 if all points in A (concatenated points in group G) are paired with the corresponding points of B (concatenated points in the associated group), else 0
+- **Cost**:  BN254g1=18000 BN254g2=18000 BLS12_381g1=15000 BLS12_381g2=15000
+- Availability: v9
+
+A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes.
+
+## ec_multi_exp g
+
+- Opcode: 0xe3 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn
+- **Cost**:  BN254g1=800 BN254g2=1800 BLS12_381g1=1400 BLS12_381g2=3500
+- Availability: v9
+
+A is a list of concatenated points, encoded and checked as described in `ec_add`. B is a list of concatenated 32 bytes scalars.
+
+## ec_subgroup_check g
+
+- Opcode: 0xe4 {uint8 curve}
+- Stack: ..., A: []byte &rarr; ..., uint64
+- whether point A is in the main prime-order subgroup of G => {0 or 1}. Program fails if A is not in G at all.
+- **Cost**:  BN254g1=50 BN254g2=11500 BLS12_381g1=5600 BLS12_381g2=7100
+- Availability: v9
+
+## ec_map_to g
+
+- Opcode: 0xe5 {uint8 curve}
+- Stack: ..., A: []byte &rarr; ..., []byte
+- maps field element A to group G
+- **Cost**:  BN254g1=1700 BN254g2=11000 BLS12_381g1=5600 BLS12_381g2=43000
+- Availability: v9
+
+BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. G1 element inputs are encoded single big-endian byte-array of length <= n.  G2 inputs are encoded as concatenated n-byte big-endian encoded integers. n == 32 for BN254 and 48 for BLS12-381. No input may exceed the modulus of the group. 
```

Comment:
```suggestion
BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. 

G1 element inputs are base field elements and G2 element inputs are quadratic field elements, with the same encoding rules as defined in `ec_add`.
In particular, no input coefficient can exceed the modulus of the group.
```

Reply:
`ec_add` takes G1 points, which are two concatenated 32 big-endian integers.  But here, we take the "base field elements" which for G1, are just single integers. So it does not seem helpful discuss the "same encoding rules as `ec_add`".

In fact, there is quite a departure from those rules, even when you consider the elements themselves. Here, at least for G1 groups, we allow short bytes sequences like 0x0108, and interpret is as 264 (256+8). We don't mind that the input is not 32 bytes long. Since `ec_add` effectively takes two such elements, it _demands_ the input is 64 bytes - two concatenated 32 byte values.

For `ec_map_to` applied to G2 fields, we can not allow such conveniences, since the base field is itself quadratic.  They must be 64 or 94 bytes (2 32s, or 2 48s).  That why I talk about "<= n" for G1 inputs, and exactly n for G2 inputs.